exec/src/coqnum.ml

```diff
@@ -35,8 +35,21 @@ let int_of_coqpositive =
     | _ -> raise (UnsupportedTerm "not a constructor of [positive]")
   in of_coqpositive_aux 0 1
 
-let int_to_coqpositive _ =
-  raise (UnsupportedTerm "To be implemented")
+let int_to_coqpositive i =
+  let rec int_to_lbool acc x =
+    if x != 0
+    then int_to_lbool ((if x mod 2 == 1 then true else false) :: acc) (x/2)
+    else acc in
+  let bool_to_coqpositivec b =
+    if b then Ind.Positive.mkConstructor "xI" else Ind.Positive.mkConstructor "xO" in
+  let rec coqpositive_of_lbool = function
+    | x1 :: x2 :: rst
+      -> let rst = coqpositive_of_lbool (x2 :: rst) in
+         let c = bool_to_coqpositivec x1 in
+         mkApp (c, Array.of_list [rst])
+    | true :: [] -> Ind.Positive.mkConstructor "xH"
+    | _ -> raise (UnsupportedTerm "not a positive integer") in
```

Comment:
This is a pretty neat implementation, but the fact that there is 3 steps makes it a bit hard to grasp I think.

While reading your implementation, I had an idea that would (maybe) be a bit easier to understand, and I am curious to have your opinion on that.

My idea is to define an auxiliary function that iterates of integer bits, like that:

```ocaml
fold_bit: int (* input int *)
       -> ('b -> bool -> 'b) (* classical function to accumulate *)
       -> ('b -> 'c) (* function to call of the last bit (known to be set) *)
       -> 'b (* initial accumulator *)
       -> 'c (* result *)
```

Then, you can use this function using a continuation passing style, with `'b` being `Constr.t -> Constr.t`, and `'c` being `Constr.t`

```ocaml
fold_bit i
         (fun f bit ->
               let c = if bit
                       then nd.Positive.mkConstructor "xI"
                       else Ind.Positive.mkConstructor "xO" in
               fun next -> mkApp c (Array.of_list [f next]))
         (fun f -> f (Ind.Positive.mkConstructor "xH"))
         (fun x -> x)
```

What do you think? @yurug?

Reply:
For the record, and because I kind of really like these kinds of challenges, I proposed a PR implemented this alternative style in #17